### DISCUSSION TITLE
CLI: How to cleanup after testing with examples?

### DISCUSSION THREAD

**Discussion by ravy** [schemathesis/schemathesis#2374] [Q&A]
### use case: 
lets say we have two APIs, `POST /tree` and `DELETE /tree/{id}`
lets say we are testing only examples using `--hypothesis-phases=explicit`
lets say i enforced order and put `DELETE /tree/{id}` after the `POST /tree` call
lets say `POST /tree` has two examples

want to delete the two trees created in POST
because after schemathesis runs need to leave the database/cache in same state as when started
what recipe can i use to solve this use case?
note1: considering just two example trees, but it could be two hundred.
note2: links are actually followed here, which is a bit weird, but the followed links only run the examples under each link

### current solution:

- using `after_call` hook, after POST call cache the tree IDs
- using `after_init_cli_run_handlers` hook, run a handler that deletes trees using IDs from cache using http `requests` (rest client)

```python
# cache of rest-paths to a list-of-IDs
id_cache = dict()

# after post call, cache the ids
@schemathesis.hook
def after_call(context, case: Case, response: GenericResponse):
    if response.status_code == 201 and case.method == "POST" and case.path == "/trees" :
        parsed = response.json()
        cache_set_id(case.path, parsed['id'])

@schemathesis.hook
def after_init_cli_run_handlers(
    context: schemathesis.hooks.HookContext,
    handlers: List[EventHandler],
    execution_context: ExecutionContext,
) -> None:
    # insert at the end of handlers list
    handlers.append(DeleteTreeHandler())
```

is there a schemathesis way of deleting a bunch of IDs that were created in examples instead of using a rest client?

**Comment by Stranger6667** [comment]
Hey!

At the moment, Schemathesis does not infer any relations among API operations, so the user has to do it manually. I'd say that using the `after_call` hook would be the way to go
